core/src/scene/stops.cpp

```diff
@@ -0,0 +1,84 @@
+#include "stops.h"
+#include "csscolorparser.hpp"
+#include "yaml-cpp/yaml.h"
+#include <cassert>
+#include <algorithm>
+
+#define RED(col) (col % 256)
+#define GRE(col) ((col >> 8) % 256)
+#define BLU(col) ((col >> 16) % 256)
+#define ALP(col) ((col >> 24) % 256)
+
+namespace Tangram {
+
+Stops::Stops(const YAML::Node& _node, bool _isColor) {
+
+    if (!_node.IsSequence()) { return; }
+
+    for (const auto& frameNode : _node) {
+        if (!frameNode.IsSequence() || frameNode.size() != 2) { continue; }
+        float key = frameNode[0].as<float>();
+        if (_isColor) {
+            // parse color from sequence or string
+            uint32_t color = 0;
+            YAML::Node colorNode = frameNode[1];
+            if (colorNode.IsScalar()) {
+                color = CSSColorParser::parse(colorNode.as<std::string>()).getInt();
+            } else if (colorNode.IsSequence() && colorNode.size() >= 3) {
+                float alpha = colorNode.size() > 3 ? colorNode[3].as<float>() : 1.f;
+                color = alpha * 255.;
+                color = (color << 8) + colorNode[2].as<float>() * 255.;
+                color = (color << 8) + colorNode[1].as<float>() * 255.;
+                color = (color << 8) + colorNode[0].as<float>() * 255.;
+            }
+            frames.emplace_back(key, color);
```

Comment:
Yes! Would like to support different interpolation schemes in the future,
but want to think more about possible syntax, for example letting user
specify color space and/or interpolation curve (cubic, ease in/out, etc.).
So just sticking with simple RGB linear for now :)

On Mon, Sep 21, 2015 at 2:28 PM, Hannes Janetzek <notifications@github.com>
wrote:

> In core/src/scene/stops.cpp
> <https://github.com/tangrams/tangram-es/pull/290#discussion_r40007069>:
>
> > +        if (!frameNode.IsSequence() || frameNode.size() != 2) { continue; }
> > +        float key = frameNode[0].as<float>();
> > +        if (_isColor) {
> > +            // parse color from sequence or string
> > +            uint32_t color = 0;
> > +            YAML::Node colorNode = frameNode[1];
> > +            if (colorNode.IsScalar()) {
> > +                color = CSSColorParser::parse(colorNode.as<std::string>()).getInt();
> > +            } else if (colorNode.IsSequence() && colorNode.size() >= 3) {
> > +                float alpha = colorNode.size() > 3 ? colorNode[3].as<float>() : 1.f;
> > +                color = alpha * 255.;
> > +                color = (color << 8) + colorNode[2].as<float>() * 255.;
> > +                color = (color << 8) + colorNode[1].as<float>() * 255.;
> > +                color = (color << 8) + colorNode[0].as<float>() * 255.;
> > +            }
> > +            frames.emplace_back(key, color);
>
> We should support other interpolations in future:
> CIE-LCh looks quite good. The 'approximation trick' should be possible to
> implement by generating more intermediate zoom steps:
> http://www.stuartdenman.com/improved-color-blending/
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/tangrams/tangram-es/pull/290/files#r40007069>.
>